Source/BoundaryConditions/PEC_Insulator.cpp

```diff
@@ -0,0 +1,593 @@
+#include <AMReX_Geometry.H>
+#include <AMReX_GpuControl.H>
+#include <AMReX_GpuLaunch.H>
+#include <AMReX_GpuQualifiers.H>
+#include <AMReX_IndexType.H>
+#include <AMReX_IntVect.H>
+#include <AMReX_MFIter.H>
+#include <AMReX_MultiFab.H>
+#include <AMReX_ParmParse.H>
+#include <AMReX_REAL.H>
+#include <AMReX_SPACE.H>
+
+namespace
+{
+    /**
+     * \brief Calculates the number of grid points the given index is beyond the
+     *        domain boundary i.e. a value of +1 means the current cell is
+     *        outside of the simulation domain by 1 cell. Note that the high
+     *        side domain boundary is between cell dom_hi and dom_hi+1 for cell
+     *        centered grids and on cell dom_hi+1 for nodal grid. This is why
+     *        (dom_hi[idim] + is_nodal[idim]) is used below.
+     *
+     * \param[in] dom_lo, dom_hi  Domain boundaries
+     * \param[in] ijk_vec         Cell coordinates
+     * \param[in] is_nodal        Whether the field of interest is nodal
+     * \param[in] idim            Dimension of interest
+     * \param[in] iside           -1 for low and +1 for high
+     *
+     * \returns number of grid points beyond the boundary
+     */
+    AMREX_GPU_DEVICE AMREX_FORCE_INLINE
+    int get_cell_count_to_boundary (amrex::IntVect const & dom_lo,
+        amrex::IntVect const & dom_hi, amrex::IntVect const & ijk_vec,
+        amrex::IntVect const & is_nodal, int idim, int iside)
+    {
+        return ((iside == -1) ? (dom_lo[idim] - ijk_vec[idim])
+                             : (ijk_vec[idim] - (dom_hi[idim] + is_nodal[idim])));
+    }
+
+
+    /**
+     * \brief At the specified grid location, apply either the PEC or insulator boundary condition if
+     *        the cell is on the boundary or in the guard cells.
+     *
+     * \param[in] icomp        component of the field being updated
+     *                         (0=x, 1=y, 2=z in Cartesian)
+     *                         (0=r, 1=theta, 2=z in RZ)
+     * \param[in] dom_lo       index value of the lower domain boundary (cell-centered)
+     * \param[in] dom_hi       index value of the higher domain boundary (cell-centered)
+     * \param[in] ijk_vec      indices along the x(i), y(j), z(k) of field Array4
+     * \param[in] n            index of the MultiFab component being updated
+     * \param[in] field        field data to be updated if (ijk) is at the boundary
+     *                         or a guard cell
+     * \param[in] E_like       whether the field behaves like E field or B field
+     * \param[in] is_nodal     staggering of the field data being updated.
+     * \param[in] is_insulator_lo Specifies whether lower boundaries are insulators
+     * \param[in] is_insulator_hi Specifies whether upper boundaries are insulators
+     * \param[in] field_lo     Specified values of the field at the lower boundaries in the insulator
+     * \param[in] field_hi     Specified values of the field at the upper boundaries in the insulator
```

Comment:
Could you add the docstrings for `set_field_lo` and `set_field_hi` as well?

Also, I think `field_lo` and `field_hi` are called `fbndry_lo` and `fbndry_hi` in the code:
```suggestion
     * \param[in] icomp        component of the field being updated
     *                         (0=x, 1=y, 2=z in Cartesian)
     *                         (0=r, 1=theta, 2=z in RZ)
     * \param[in] dom_lo       index value of the lower domain boundary (cell-centered)
     * \param[in] dom_hi       index value of the higher domain boundary (cell-centered)
     * \param[in] ijk_vec      indices along the x(i), y(j), z(k) of field Array4
     * \param[in] n            index of the MultiFab component being updated
     * \param[in] field        field data to be updated if (ijk) is at the boundary
     *                         or a guard cell
     * \param[in] E_like       whether the field behaves like E field or B field
     * \param[in] is_nodal     staggering of the field data being updated.
     * \param[in] is_insulator_lo Specifies whether lower boundaries are insulators
     * \param[in] is_insulator_hi Specifies whether upper boundaries are insulators
     * \param[in] fbndry_lo     Specified values of the field at the lower boundaries in the insulator
     * \param[in] fbndry_hi     Specified values of the field at the upper boundaries in the insulator
```